File: scripts/job_apply_policy/campaigns.py
"""Campaign lifecycle policy operations."""
from __future__ import annotations

from datetime import datetime, timedelta
from typing import Any

from .model import (
    CAMPAIGN_FIELDS, CAMPAIGN_STATUSES, MAX_APPLICATIONS, MAX_DURATION,
    SCHEMA_VERSION, PolicyError, _closed, _fingerprint, _new_reference,
    _reference, _rule, _sensitive, format_time, parse_time, utc_now,
)
from .storage import _private_dir, _read_json

class CampaignMixin:
    def _validate_campaign(self, campaign: dict[str, Any]) -> dict[str, Any]:
        _closed(campaign, CAMPAIGN_FIELDS, "campaign")
        if campaign["schemaVersion"] != SCHEMA_VERSION:
            raise PolicyError("campaign schema version is unsupported")
        _reference(campaign["campaignId"], "campaign", "campaignId")
        if campaign["mode"] != "auto_submit":
            raise PolicyError("campaign mode is invalid")
        if campaign["status"] not in CAMPAIGN_STATUSES:
            raise PolicyError("campaign status is invalid")
        created = parse_time(campaign["createdAt"])
        expires = parse_time(campaign["expiresAt"])
        if expires <= created or expires - created > MAX_DURATION:
            raise PolicyError("campaign duration is invalid")
        maximum = campaign["maxApplications"]
        if not isinstance(maximum, int) or isinstance(maximum, bool) or not 1 <= maximum <= MAX_APPLICATIONS:
            raise PolicyError("campaign application limit is invalid")
        rules = campaign["applicationRules"]
        if not isinstance(rules, list) or not rules:
            raise PolicyError("campaign requires application rules")
        normalized_rules = [_rule(rule) for rule in rules]
        refs = [rule["applicationRef"] for rule in normalized_rules]
        if len(refs) != len(set(refs)):
            raise PolicyError("campaign application rules contain duplicates")
        _fingerprint(campaign["resumeRevision"], "resumeRevision")
        _fingerprint(
            campaign["confirmationAuthorityRevision"],
            "confirmationAuthorityRevision",
        )
        allowlist = campaign["sensitiveAllowlist"]
        if not isinstance(allowlist, list):
            raise PolicyError("sensitiveAllowlist must be a list")
        normalized_allowlist = [_sensitive(item) for item in allowlist]
        answer_refs = [item["answerRef"] for item in normalized_allowlist]
        if len(answer_refs) != len(set(answer_refs)):
            raise PolicyError("sensitiveAllowlist contains duplicates")
        parse_time(campaign["riskAcknowledgedAt"])
        if not isinstance(campaign["killSwitch"], bool):
            raise PolicyError("campaign kill switch is invalid")
        if campaign["killSwitchAt"] is not None:
            parse_time(campaign["killSwitchAt"])
        if campaign["killSwitch"] != (campaign["status"] == "killed"):
            raise PolicyError("campaign kill switch state is inconsistent")
        if campaign["killSwitch"] != (campaign["killSwitchAt"] is not None):
            raise PolicyError("campaign kill switch timestamp is inconsistent")
        reserved = campaign["reservedApplications"]
        if not isinstance(reserved, int) or isinstance(reserved, bool) or not 0 <= reserved <= maximum:
            raise PolicyError("campaign reservation count is invalid")
        return campaign

    def load_campaign(self) -> dict[str, Any]:
        if not self.campaign_path.exists():
            raise PolicyError("campaign does not exist")
        return self._validate_campaign(_read_json(self.campaign_path, "campaign"))

    def _campaign_mode(self, campaign: dict[str, Any], now: datetime) -> tuple[str, str]:
        if campaign["killSwitch"]:
            return "review_only", "kill_switch"
        if campaign["status"] != "active":
            return "review_only", campaign["status"]
        if now >= parse_time(campaign["expiresAt"]):
            return "review_only", "expired"
        return "auto_submit", "active_campaign"

    def decision(self, now: datetime | None = None) -> dict[str, Any]:
        checked_at = now or utc_now()
        try:
            campaign = self.load_campaign()
            mode, reason = self._campaign_mode(campaign, checked_at)
            return {
                "mode": mode,
                "reason": reason,
                "campaignId": campaign["campaignId"] if mode == "auto_submit" else None,
            }
        except (PolicyError, OSError):
            return {"mode": "review_only", "reason": "policy_unavailable", "campaignId": None}

    def activate(self, incoming: dict[str, Any], now: datetime | None = None) -> dict[str, Any]:
        current_time = now or utc_now()
        allowed = {
            "riskAcknowledged",
            "applicationRules",
            "resumeRevision",
            "sensitiveAllowlist",
            "confirmationAuthorityRevision",
            "maxApplications",
            "durationSeconds",
        }
        if set(incoming) - allowed:
            raise PolicyError("campaign input contains unsupported fields")
        if incoming.get("riskAcknowledged") is not True:
            raise PolicyError("explicit risk acknowledgement is required")
        rules = incoming.get("applicationRules")
        if not isinstance(rules, list) or not rules:
            raise PolicyError("campaign requires application rules")
        normalized_rules = [_rule(rule) for rule in rules]
        maximum = incoming.get("maxApplications", MAX_APPLICATIONS)
        duration_seconds = incoming.get("durationSeconds", int(MAX_DURATION.total_seconds()))
        if not isinstance(duration_seconds, int) or isinstance(duration_seconds, bool) or not 1 <= duration_seconds <= int(MAX_DURATION.total_seconds()):
            raise PolicyError("campaign duration is invalid")
        campaign = {
            "schemaVersion": SCHEMA_VERSION,
            "campaignId": _new_reference("campaign"),
            "mode": "auto_submit",
            "status": "active",
            "createdAt": format_time(current_time),
            "expiresAt": format_time(current_time + timedelta(seconds=duration_seconds)),
            "maxApplications": maximum,
            "applicationRules": normalized_rules,
            "resumeRevision": _fingerprint(incoming.get("resumeRevision"), "resumeRevision"),
            "sensitiveAllowlist": [
                _sensitive(item) for item in incoming.get("sensitiveAllowlist", [])
            ],
            "confirmationAuthorityRevision": _fingerprint(
                incoming.get("confirmationAuthorityRevision"),
                "confirmationAuthorityRevision",
            ),
            "riskAcknowledgedAt": format_time(current_time),
            "killSwitch": False,
            "killSwitchAt": None,
            "reservedApplications": 0,
        }
        self._validate_campaign(campaign)
        with self._lock():
            if self.campaign_path.exists():
                previous = self.load_campaign()
                previous_mode, _ = self._campaign_mode(previous, current_time)
                if previous_mode == "auto_submit" or previous["killSwitch"]:
                    raise PolicyError("an active or killed campaign cannot be replaced")
                _private_dir(self.archive_dir)
                archive_path = self.archive_dir / f"{previous['campaignId'].split(':', 1)[1]}.json"
                if archive_path.exists():
                    archived = self._validate_campaign(
                        _read_json(archive_path, "campaign archive")
                    )
                    if archived != previous:
                        raise PolicyError("campaign archive already exists")
                else:
                    self._write_json(archive_path, previous)
            self._write_json(self.campaign_path, campaign)
        return campaign

    def _load_campaign_by_id(self, campaign_id: str) -> dict[str, Any]:
        _reference(campaign_id, "campaign", "campaignId")
        current = self.load_campaign()
        if current["campaignId"] == campaign_id:
            return current
        archive_path = self.archive_dir / f"{campaign_id.split(':', 1)[1]}.json"
        if not archive_path.exists():
            raise PolicyError("campaign does not exist")
        archived = self._validate_campaign(
            _read_json(archive_path, "campaign archive")
        )
        if archived["campaignId"] != campaign_id:
            raise PolicyError("campaign archive does not match")
        return archived

File: src/final-action-policy/campaigns.ts
import { join } from 'node:path';
import { PolicyRepository, exists, readDocument } from './repository.js';
import { check, closed, fingerprint, formatTime, MAX_DURATION, object, parseRule, parseSensitive, parseTime, reference, canonical, unique } from './model.js';
import type { Campaign } from './model.js';

export function parseCampaign(value: unknown): Campaign {
  const campaign = object(value, 'campaign');
  closed(campaign, ['schemaVersion', 'campaignId', 'mode', 'status', 'createdAt', 'expiresAt', 'maxApplications', 'applicationRules', 'resumeRevision', 'sensitiveAllowlist', 'confirmationAuthorityRevision', 'riskAcknowledgedAt', 'killSwitch', 'killSwitchAt', 'reservedApplications'], 'campaign');
  check(campaign.schemaVersion === 1, 'campaign schema version is unsupported');
  reference(campaign.campaignId, 'campaign', 'campaignId');
  check(campaign.mode === 'auto_submit', 'campaign mode is invalid');
  check(typeof campaign.status === 'string' && ['active', 'revoked', 'killed', 'expired'].includes(campaign.status), 'campaign status is invalid');
  const created = parseTime(campaign.createdAt), expires = parseTime(campaign.expiresAt);
  check(expires > created && expires - created <= MAX_DURATION, 'campaign duration is invalid');
  check(Number.isSafeInteger(campaign.maxApplications) && Number(campaign.maxApplications) >= 1 && Number(campaign.maxApplications) <= 10, 'campaign application limit is invalid');
  check(Array.isArray(campaign.applicationRules) && campaign.applicationRules.length > 0, 'campaign requires application rules');
  const rules = campaign.applicationRules.map(parseRule);
  unique(rules, rule => rule.applicationRef, 'campaign application rules contain duplicates');
  fingerprint(campaign.resumeRevision, 'resumeRevision');
  fingerprint(campaign.confirmationAuthorityRevision, 'confirmationAuthorityRevision');
  check(Array.isArray(campaign.sensitiveAllowlist), 'sensitiveAllowlist must be a list');
  const allowlist = campaign.sensitiveAllowlist.map(parseSensitive);
  unique(allowlist, item => item.answerRef, 'sensitiveAllowlist contains duplicates');
  parseTime(campaign.riskAcknowledgedAt);
  check(typeof campaign.killSwitch === 'boolean', 'campaign kill switch is invalid');
  if (campaign.killSwitchAt !== null) parseTime(campaign.killSwitchAt);
  check(campaign.killSwitch === (campaign.status === 'killed'), 'campaign kill switch state is inconsistent');
  check(campaign.killSwitch === (campaign.killSwitchAt !== null), 'campaign kill switch timestamp is inconsistent');
  check(Number.isSafeInteger(campaign.reservedApplications) && Number(campaign.reservedApplications) >= 0 && Number(campaign.reservedApplications) <= Number(campaign.maxApplications), 'campaign reservation count is invalid');
  return structuredClone(campaign) as unknown as Campaign;
}
export function campaignMode(campaign: Campaign, now: Date): { mode: 'review_only' | 'auto_submit'; reason: string } {
  if (campaign.killSwitch) return { mode: 'review_only', reason: 'kill_switch' };
  if (campaign.status !== 'active') return { mode: 'review_only', reason: campaign.status };
  if (now.getTime() >= parseTime(campaign.expiresAt)) return { mode: 'review_only', reason: 'expired' };
  return { mode: 'auto_submit', reason: 'active_campaign' };
}
export class CampaignPolicy extends PolicyRepository {
  protected async loadCampaign(): Promise<Campaign> {
    check(await exists(this.campaignPath), 'campaign does not exist');
    return parseCampaign(await readDocument(this.campaignPath, 'campaign'));
  }
  protected async loadCampaignById(campaignId: string): Promise<Campaign> {
    reference(campaignId, 'campaign', 'campaignId');
    const current = await this.loadCampaign();
    if (current.campaignId === campaignId) return current;
    const path = join(this.archiveDir, `${campaignId.split(':')[1]}.json`);
    check(await exists(path), 'campaign does not exist');
    const archived = parseCampaign(await readDocument(path, 'campaign archive'));
    check(archived.campaignId === campaignId, 'campaign archive does not match');
    return archived;
  }
  async status(): Promise<{ mode: string; reason: string; campaignId: string | null }> {
    // Python status is a pure observation, never a grant and never a writer.
    try {
      const campaign = await this.loadCampaign();
      const decision = campaignMode(campaign, this.clock());
      return { ...decision, campaignId: decision.mode === 'auto_submit' ? campaign.campaignId : null };
    } catch { return { mode: 'review_only', reason: 'policy_unavailable', campaignId: null }; }
  }
  async activate(value: unknown): Promise<Campaign> {
    const incoming = object(value, 'campaign input');
    const now = this.clock();
    const allowed = ['riskAcknowledged', 'applicationRules', 'resumeRevision', 'sensitiveAllowlist', 'confirmationAuthorityRevision', 'maxApplications', 'durationSeconds'];
    check(Object.keys(incoming).every(key => allowed.includes(key)), 'campaign input contains unsupported fields');
    check(incoming.riskAcknowledged === true, 'explicit risk acknowledgement is required');
    check(Array.isArray(incoming.applicationRules) && incoming.applicationRules.length > 0, 'campaign requires application rules');
    const rules = incoming.applicationRules.map(parseRule);
    const duration = Object.hasOwn(incoming, 'durationSeconds') ? incoming.durationSeconds : MAX_DURATION / 1000;
    check(Number.isSafeInteger(duration) && Number(duration) >= 1 && Number(duration) <= MAX_DURATION / 1000, 'campaign duration is invalid');
    const campaign = parseCampaign({ schemaVersion: 1, campaignId: this.newReference('campaign'), mode: 'auto_submit', status: 'active',
      createdAt: formatTime(now), expiresAt: formatTime(new Date(now.getTime() + Number(duration) * 1000)),
      maxApplications: Object.hasOwn(incoming, 'maxApplications') ? incoming.maxApplications : 10,
      applicationRules: rules, resumeRevision: fingerprint(incoming.resumeRevision, 'resumeRevision'),
      sensitiveAllowlist: Object.hasOwn(incoming, 'sensitiveAllowlist') ? incoming.sensitiveAllowlist : [],
      confirmationAuthorityRevision: fingerprint(incoming.confirmationAuthorityRevision, 'confirmationAuthorityRevision'),
      riskAcknowledgedAt: formatTime(now), killSwitch: false, killSwitchAt: null, reservedApplications: 0 });
    return this.locked(async () => {
      if (await exists(this.campaignPath)) {
        const previous = await this.loadCampaign();
        check(campaignMode(previous, now).mode !== 'auto_submit' && !previous.killSwitch, 'an active or killed campaign cannot be replaced');
        await this.privateDirectory(this.archiveDir);
        const path = join(this.archiveDir, `${previous.campaignId.split(':')[1]}.json`);
        if (await exists(path)) {
          const archived = parseCampaign(await readDocument(path, 'campaign archive'));
          check(canonical(archived) === canonical(previous), 'campaign archive already exists');
        } else await this.writeDocument(path, previous);
      }
      await this.writeDocument(this.campaignPath, campaign);
      return campaign;
    });
  }
  private async stop(status: 'killed' | 'revoked', kill: boolean): Promise<{ mode: string; reason: string; campaignId: string }> {
    const now = this.clock();
    return this.locked(async () => {
      const campaign = await this.loadCampaign();
      check(!campaign.killSwitch || kill, 'kill switch cannot be cleared by revocation');
      campaign.status = status;
      campaign.killSwitch = kill;
      campaign.killSwitchAt = kill ? formatTime(now) : null;
      await this.writeDocument(this.campaignPath, parseCampaign(campaign));
      return { mode: 'review_only', reason: status, campaignId: campaign.campaignId };
    });
  }
  kill(): Promise<{ mode: string; reason: string; campaignId: string }> { return this.stop('killed', true); }
  revoke(): Promise<{ mode: string; reason: string; campaignId: string }> { return this.stop('revoked', false); }
}
